packages/discussions/src/utils/channels/can-create-channel.ts

```diff
@@ -0,0 +1,192 @@
+import { IGroup, IUser } from "@esri/arcgis-rest-types";
+import {
+  AclCategory,
+  IChannel,
+  IChannelAclPermissionDefinition,
+  IDiscussionsUser,
+  SharingAccess,
+} from "../../types";
+import { isOrgAdmin } from "../platform";
+
+const ALLOWED_USER_GROUP_ROLES = Object.freeze(["owner", "admin", "member"]);
+
+type ILegacyChannelPermissions = Pick<IChannel, "access" | "groups" | "orgs">;
+
+type PermissionsDefinitionsByAclCategoryMap = {
+  [key in AclCategory]?: IChannelAclPermissionDefinition[];
+};
+
+export function canCreateChannel(
+  channel: IChannel,
+  user: IDiscussionsUser
+): boolean {
+  const { channelAcl, access, groups, orgs } = channel;
+
+  if (channelAcl) {
+    return isAuthorizedToCreateByChannelAcl(user, channelAcl);
+  }
+
+  return isAuthorizedToCreateByLegacyPermissions(user, {
+    access,
+    groups,
+    orgs,
+  });
+}
+
+function isAuthorizedToCreateByChannelAcl(
+  user: IDiscussionsUser,
+  channelAcl: IChannelAclPermissionDefinition[]
+): boolean {
+  const { username } = user;
+
+  if (username === null || channelAcl.length === 0) {
+    return false;
+  }
+
+  const permissions = mapByCategory(channelAcl);
+
+  return (
+    canAllowAnonymous(user, permissions[AclCategory.ANONYMOUS_USER]) &&
+    canAllowAuthenticated(user, permissions[AclCategory.AUTHENTICATED_USER]) &&
+    canAllowGroups(user, permissions[AclCategory.GROUP]) &&
+    canAllowOrgs(user, permissions[AclCategory.ORG]) &&
+    canAllowUsers(user, permissions[AclCategory.USER])
+  );
+}
+
+function mapByCategory(channelAcl: IChannelAclPermissionDefinition[]) {
+  return channelAcl.reduce((accum, permission) => {
+    const { category } = permission;
+
+    accum[category]?.push(permission) || (accum[category] = [permission]);
+    return accum;
+  }, {} as PermissionsDefinitionsByAclCategoryMap);
+}
+
+function canAllowAnonymous(
+  user: IDiscussionsUser,
+  anonPermissions?: IChannelAclPermissionDefinition[]
+) {
+  if (!anonPermissions) {
+    return true;
+  }
+  return isOrgAdmin(user);
+}
+
+function canAllowAuthenticated(
+  user: IDiscussionsUser,
+  authenticatedPermissions?: IChannelAclPermissionDefinition[]
+) {
+  if (!authenticatedPermissions) {
+    return true;
+  }
+  return isOrgAdmin(user);
+}
+
+function canAllowGroups(
+  user: IDiscussionsUser,
+  groupPermissions?: IChannelAclPermissionDefinition[]
+) {
+  if (!groupPermissions) {
+    return true;
+  }
+  return isMemberOfAllChannelGroups(user.groups, groupPermissions);
```

Comment:
membership check is good, but you also have to check the group's typeKeyword array for existence of `cannotDiscuss`.  See https://devtopia.esri.com/dc/hub-discussions/blob/main/src/common/utils/validate-all-groups-can-discuss.ts#L52-L60.

Reply:
👍